patches/0132-ffmpeg-raisr-filter.patch

```diff
@@ -0,0 +1,443 @@
++    AVFilterLink *inlink0 = outlink->src->inputs[0];
++
++    outlink->w = inlink0->w * raisr->ratio;
++    outlink->h = inlink0->h * raisr->ratio;
++
++    return 0;
++}
++
++static int filter_frame(AVFilterLink *inlink, AVFrame *in)
++{
++    AVFilterContext *ctx = inlink->dst;
++    RaisrContext *raisr = ctx->priv;
++    AVFilterLink *outlink = ctx->outputs[0];
++    AVFrame *out;
++    RNLERRORTYPE ret;
++    VideoDataType vdt_in[3] = { 0 };
++    VideoDataType vdt_out[3] = { 0 };
++
++    av_log(ctx, AV_LOG_VERBOSE, "Frame\n");
++
++    // Request a picture buffer - must be released with. This must be unreferenced with
++    // avfilter_unref_buffer when you are finished with it
++    out = ff_get_video_buffer(outlink, outlink->w, outlink->h);
++    if (!out)
++    {
++        // Unable to get a picture buffer.
++        // Delete the input buffer and return
++        av_frame_free(&in);
++        return AVERROR(ENOMEM);
++    }
++    av_log(ctx, AV_LOG_VERBOSE, "Got Frame %dx%d\n", outlink->w, outlink->h);
++
++    // Copy only "metadata" fields from src to dst.
++    // Metadata for the purpose of this function are those fields that do not affect
++    // the data layout in the buffers.
++    av_frame_copy_props(out, in);
++    av_log(ctx, AV_LOG_VERBOSE, "Copied props \n");
++
++    // For each plane
++    for (int p = 0; p < raisr->nb_planes; p++)
++    {
++        // get the plane data
++        struct plane_info *plane = &raisr->inplanes[p];
++
++        // make sure the input data is valid
++        av_assert1(in->data[p]);
++
++        // get a pointer to the out plane data
++        av_assert1(out->data[p]);
++
++        // fill in the input video data type structure
++        vdt_in[p].pData = in->data[p];
++        vdt_in[p].width = plane->width;
++        vdt_in[p].height = plane->height;
++        vdt_in[p].step = in->linesize[p];
++
++        // fill in the output video data type structure
++        vdt_out[p].pData = out->data[p];
++        vdt_out[p].width = plane->width * raisr->ratio;
++        vdt_out[p].height = plane->height * raisr->ratio;
++        vdt_out[p].step = out->linesize[p];
++    }
++    if (raisr->framecount == 0)
++    {
++        // Process the planes
++        ret = RNLHandler_SetRes(
++            &vdt_in[0],
++            &vdt_in[1],
++            &vdt_in[2],
++            &vdt_out[0],
++            &vdt_out[1],
++            &vdt_out[2]);
++
++        if (ret != RNLErrorNone)
++        {
++            av_log(ctx, AV_LOG_INFO, "RNLHandler_SetRes error\n");
++            return AVERROR(ENOMEM);
++        }
++    }
++
++    // Process the planes
++    ret = RNLHandler_Process(
++        &vdt_in[0],
++        &vdt_in[1],
++        &vdt_in[2],
++        &vdt_out[0],
++        &vdt_out[1],
++        &vdt_out[2],
++        raisr->blending);
++
++    if (ret != RNLErrorNone)
++    {
++        av_log(ctx, AV_LOG_INFO, "RNLHandler_Process error\n");
++        return AVERROR(ENOMEM);
```

Comment:
in and out are not freed in the error path.

Reply:
Done